{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### filter conditions:\n",
    "#### 1. only limit IP total number \n",
    "####    1.1 input (refCount, altCount >=2 and refCount+altCount >=10)\n",
    "####    1.2 ip (refCount+altCount >=10)\n",
    "#### 2. IP/Input >=2\n",
    "#### 3. within m6A peak (remove m6Am)\n",
    "#### filter order\n",
    "#### remove SNPs which don't meet minimum needs\n",
    "#### Hypothesis test\n",
    "#### process 2,3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CRR055553 736\n",
      "CRR073016 595\n",
      "CRR073020 298\n",
      "CRR042292 1340\n",
      "CRR042294 1112\n",
      "SRR8209849 2345\n",
      "CRR073018 869\n",
      "SRR8209863 1135\n",
      "CRR042282 1543\n",
      "SRR8209857 1429\n",
      "SRR8209875 2119\n",
      "SRR8209837 3486\n",
      "SRR8209853 1794\n",
      "SRR8209855 2896\n",
      "SRR8209867 1274\n",
      "CRR042284 3797\n",
      "SRR8209859 985\n",
      "CRR055527 3119\n",
      "SRR8209865 1407\n",
      "liver_IP_1 1789\n",
      "SRR8209871 5459\n",
      "SRR8209841 12936\n",
      "muscle_IP_5 5675\n",
      "heart_IP_2 3898\n",
      "liver_IP_3 3094\n",
      "muscle_IP_4 3730\n",
      "lung_IP_4 4979\n",
      "SRR8209839 16775\n",
      "brain_IP_2 5761\n",
      "brain_IP_3 6592\n",
      "SRR8209843 754\n",
      "SRR8209877 770\n",
      "lung_IP_5 2430\n",
      "SRR8209873 890\n",
      "SRR8209847 7463\n",
      "kidney_IP_4 5619\n",
      "SRR8209879 2747\n",
      "CRR055533 353\n",
      "SRR8209861 4615\n",
      "brain_IP_1 6826\n",
      "CRR042286 1446\n",
      "CRR042290 1259\n",
      "heart_IP_3 3638\n",
      "kidney_IP_2 7777\n",
      "SRR8209869 3972\n",
      "SRR8209845 11176\n",
      "heart_IP_1 2240\n",
      "liver_IP_2 3381\n",
      "kidney_IP_3 4923\n",
      "CRR042296 1306\n",
      "CRR042310 957\n",
      "CRR042320 497\n",
      "placenta_IP_6 4106\n",
      "CRR055555 390\n",
      "CRR042300 1033\n",
      "stomach_IP_4 3308\n",
      "CRR042280 1744\n",
      "CRR055539 393\n",
      "CRR042316 1635\n",
      "CRR042302 670\n",
      "SRR8209881 6233\n",
      "CRR055543 847\n",
      "CRR055547 760\n",
      "stomach_IP_5 2401\n",
      "CRR055551 565\n",
      "placenta_IP_4 5789\n",
      "CRR042278 4492\n",
      "CRR042304 4467\n",
      "CRR055549 486\n",
      "SRR8209883 2229\n",
      "CRR055529 2215\n",
      "CRR042308 888\n",
      "CRR042298 642\n",
      "CRR055557 313\n",
      "SRR8209885 10057\n",
      "CRR042306 1324\n",
      "CRR055559 247\n",
      "CRR055561 547\n",
      "CRR055563 307\n",
      "CRR042314 930\n",
      "CRR042312 1027\n",
      "CRR042318 2889\n",
      "CRR055537 683\n",
      "CRR042288 534\n",
      "CRR055525 4135\n",
      "placenta_IP_2 6160\n",
      "CRR055536 776\n",
      "CRR055545 1086\n",
      "CRR055541 419\n",
      "CRR055531 1306\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "from multiprocessing import Pool\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "\n",
    "#############################################################\n",
    "os.chdir(\"/Charles/project/ASm6A/by_ind/count/\")\n",
    "result_dir = \"/Charles/project/ASm6A/Hypothesis_test/ASm6A/\"\n",
    "os.system(\"mkdir -p %s\" % result_dir)\n",
    "#############################################################\n",
    "\n",
    "\n",
    "def generate_totalNum_dict():\n",
    "    df_map = pd.read_table(\n",
    "        \"/home/galaxy/project/alleleSpecific_analysis/data/sample_info/total_Input_vs_IP.txt\"\n",
    "    )\n",
    "    map_dict = dict(zip(df_map['IP'], df_map['Input']))\n",
    "    summary_file = \"/Charles/project/ASm6A/bam_count/AllSamples.bam_stat.txt\"\n",
    "    df_stat = pd.read_table(summary_file, header=None)\n",
    "    number_dict = dict(\n",
    "        zip(df_stat.iloc[:, 0].tolist(), df_stat.iloc[:, 1].tolist()))\n",
    "    totalNum_dict = {}\n",
    "    for sample in number_dict.keys():\n",
    "        if sample in map_dict:\n",
    "            totalNum_dict[sample] = [\n",
    "                number_dict[sample], map_dict[sample],\n",
    "                number_dict[map_dict[sample]]\n",
    "            ]\n",
    "    return totalNum_dict\n",
    "\n",
    "\n",
    "def preprocess_df(ip_prefix):\n",
    "    input_prefix = totalNum_dict[ip_prefix][1]\n",
    "    ip_file, input_file = \"ip_count/%s.readcounts.txt\" % ip_prefix, \"input_count/%s.readcounts.txt\" % input_prefix\n",
    "    df_ip, df_input = pd.read_table(ip_file), pd.read_table(input_file)\n",
    "    df_results = []\n",
    "    totalNum_list = [totalNum_dict[ip_prefix][0], totalNum_dict[ip_prefix][2]]\n",
    "    treat_list, df_list = ['ip', 'input'], [df_ip, df_input]\n",
    "    for i in range(len(df_list)):\n",
    "        df, treat, total_num = df_list[i], treat_list[i], totalNum_list[i]\n",
    "        df = df[[\n",
    "            'contig', 'position', 'refAllele', 'altAllele', 'refCount',\n",
    "            'altCount'\n",
    "        ]]\n",
    "        #######################################################\n",
    "        if i == 0:  # ip sample: only restrict total number\n",
    "            df = df[(df['refCount'] + df['altCount']) >= 10]\n",
    "        else:\n",
    "            df = df[(df['refCount'] >= 2) & (df['altCount'] >= 2)]\n",
    "            df = df[(df['refCount'] + df['altCount']) >= 10]\n",
    "        #####################################################################\n",
    "        df['refRPKM_%s' % treat] = (df['refCount'] / total_num) * 1000000000\n",
    "        df['altRPKM_%s' % treat] = (df['altCount'] / total_num) * 1000000000\n",
    "        df_results.append(df)\n",
    "    return df_results[0], df_results[1]\n",
    "\n",
    "\n",
    "def fisher_exact_test_each(ip_prefix):\n",
    "    result_file = os.path.join(result_dir, ip_prefix + \".txt\")\n",
    "    df_ip, df_input = preprocess_df(ip_prefix)\n",
    "    df = df_ip.merge(df_input,\n",
    "                     on=['contig', 'position', 'refAllele', 'altAllele'],\n",
    "                     how='left').dropna(how=\"any\")\n",
    "    print(ip_prefix, len(df))\n",
    "    df['refRPKM_ratio'] = df['refRPKM_ip'] / df['refRPKM_input']\n",
    "    df['altRPKM_ratio'] = df['altRPKM_ip'] / df['altRPKM_input']\n",
    "    df['allelicRatio'] = df['refRPKM_ratio'] / (df['altRPKM_ratio'] +\n",
    "                                                df['refRPKM_ratio'])\n",
    "    pvalue_list, odds_list = [], []\n",
    "    for i, j in df.iterrows():\n",
    "        a, b = j['refRPKM_ip'] + 0.5, j['altRPKM_ip'] + 0.5\n",
    "        c, d = j['refRPKM_input'] + 0.5, j['altRPKM_input'] + 0.5\n",
    "        oddsratio, pvalue = stats.fisher_exact([[a, b], [c, d]])\n",
    "        pvalue_list.append(pvalue)\n",
    "        odds_list.append(oddsratio)\n",
    "    qvalue_list = list(fdrcorrection(pvalue_list, method=\"indep\")[1])  # fdr_bh\n",
    "    df['pvalue'], df['oddsratio'], df[\n",
    "        'qvalue'] = pvalue_list, odds_list, qvalue_list\n",
    "    df_sub = df[[\n",
    "        'contig', 'position', 'refAllele', 'altAllele', 'refRPKM_ratio',\n",
    "        'altRPKM_ratio', 'allelicRatio', 'pvalue', 'oddsratio', 'qvalue'\n",
    "    ]]\n",
    "    df_sub.to_csv(result_file, sep=\"\\t\", index=False)\n",
    "\n",
    "\n",
    "totalNum_dict = generate_totalNum_dict()\n",
    "prefix_list = totalNum_dict.keys()\n",
    "# print(prefix_list)\n",
    "pool = Pool(processes=30)\n",
    "for ip_prefix in prefix_list:\n",
    "    #     fisher_exact_test_each(ip_prefix)\n",
    "    pool.apply_async(fisher_exact_test_each, (ip_prefix, ))\n",
    "pool.close()\n",
    "pool.join()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### continue to filter\n",
    "##### 1. sig (qvalue < 0.05)\n",
    "#####    1.1 within m6A peak\n",
    "#####           1.1.1 ip/input >= 2 \n",
    "\n",
    "######\n",
    "##### 2. unsig (qvalue < 0.05)\n",
    "#####     2.1 within m6A peak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CRR055533.txt 333 20\n",
      "SRR8209881.txt 4878 1355\n",
      "CRR042298.txt 562 80\n",
      "CRR042290.txt 1071 188\n",
      "CRR042304.txt 3599 868\n",
      "SRR8209869.txt 3166 806\n",
      "SRR8209875.txt 1722 397\n",
      "SRR8209841.txt 8776 4160\n",
      "CRR055537.txt 606 77\n",
      "CRR042296.txt 1147 159\n",
      "SRR8209853.txt 1531 263\n",
      "SRR8209845.txt 7557 3619\n",
      "CRR055561.txt 504 43\n",
      "SRR8209849.txt 1962 383\n",
      "SRR8209857.txt 1245 184\n",
      "SRR8209883.txt 1958 271\n",
      "CRR055531.txt 1167 139\n",
      "CRR055527.txt 2606 513\n",
      "CRR055536.txt 697 79\n",
      "CRR055555.txt 364 26\n",
      "heart_IP_2.txt 3098 800\n",
      "lung_IP_4.txt 3783 1196\n",
      "stomach_IP_5.txt 1936 465\n",
      "heart_IP_1.txt 1817 423\n",
      "muscle_IP_4.txt 2962 768\n",
      "liver_IP_3.txt 2545 549\n",
      "SRR8209837.txt 2796 690\n",
      "SRR8209839.txt 11012 5763\n",
      "CRR055529.txt 1912 303\n",
      "heart_IP_3.txt 2845 793\n",
      "brain_IP_2.txt 4143 1618\n",
      "CRR042302.txt 585 85\n",
      "CRR055543.txt 772 75\n",
      "CRR055539.txt 352 41\n",
      "CRR055525.txt 3326 809\n",
      "muscle_IP_5.txt 4358 1317\n",
      "brain_IP_3.txt 4784 1808\n",
      "SRR8209859.txt 889 96\n",
      "CRR042308.txt 780 108\n",
      "CRR073016.txt 526 69\n",
      "CRR042282.txt 1301 242\n",
      "SRR8209867.txt 1068 206\n",
      "CRR055545.txt 961 125\n",
      "SRR8209861.txt 3688 927\n",
      "CRR055553.txt 639 97\n",
      "SRR8209877.txt 678 92\n",
      "brain_IP_1.txt 4919 1907\n",
      "CRR042306.txt 1136 188\n",
      "CRR042288.txt 492 42\n",
      "placenta_IP_2.txt 4727 1433\n",
      "CRR055549.txt 442 44\n",
      "SRR8209871.txt 4177 1282\n",
      "placenta_IP_6.txt 3264 842\n",
      "CRR042320.txt 452 45\n",
      "SRR8209863.txt 997 138\n",
      "placenta_IP_4.txt 4291 1498\n",
      "CRR042316.txt 1415 220\n",
      "CRR073018.txt 763 106\n",
      "CRR042312.txt 897 130\n",
      "CRR055559.txt 226 21\n",
      "lung_IP_5.txt 1952 478\n",
      "CRR055547.txt 662 98\n",
      "liver_IP_1.txt 1445 344\n",
      "liver_IP_2.txt 2738 643\n",
      "SRR8209865.txt 1216 191\n",
      "CRR042286.txt 1213 233\n",
      "SRR8209843.txt 660 94\n",
      "kidney_IP_4.txt 4134 1485\n",
      "CRR055563.txt 288 19\n",
      "SRR8209847.txt 5683 1780\n",
      "CRR042314.txt 812 118\n",
      "CRR042278.txt 3541 951\n",
      "kidney_IP_2.txt 5662 2115\n",
      "SRR8209885.txt 7192 2865\n",
      "CRR042280.txt 1463 281\n",
      "CRR042318.txt 2331 558\n",
      "SRR8209879.txt 2287 460\n",
      "CRR073020.txt 281 17\n",
      "CRR042310.txt 836 121\n",
      "CRR055557.txt 293 20\n",
      "stomach_IP_4.txt 2587 721\n",
      "CRR042292.txt 1125 215\n",
      "CRR042300.txt 908 125\n",
      "CRR042284.txt 2858 939\n",
      "CRR042294.txt 964 148\n",
      "kidney_IP_3.txt 3659 1264\n",
      "CRR055551.txt 518 47\n",
      "SRR8209873.txt 801 89\n",
      "SRR8209855.txt 2415 481\n",
      "CRR055541.txt 385 34\n"
     ]
    }
   ],
   "source": [
    "def select_sig_and_unsig(in_file, data_type): # data_type: sig or unsig    \n",
    "    df = pd.read_table(in_file)\n",
    "    #############################\n",
    "    if data_type == \"sig\":\n",
    "        df = df[(df['qvalue']<0.05)] ##### 不同的过滤条件，用于不同的分析内容\n",
    "        # df = df[(df['qvalue']<0.1)]\n",
    "#         df = df[(df['pvalue']<0.05)]\n",
    "    elif data_type == \"unsig\":\n",
    "        df = df[(df['qvalue']>=0.05)] ####\n",
    "        # df = df[(df['qvalue']>=0.1)]\n",
    "#         df = df[(df['pvalue']>=0.05)]\n",
    "    #############################\n",
    "    df = df.round(2)\n",
    "    df['mark'] = np.where(df['altRPKM_ratio'] > df['refRPKM_ratio'], \"alt\", \"ref\")\n",
    "    df['start'] = df['position'] - 1\n",
    "    df['term'] = df['refAllele'] + \">\" + df['altAllele'] + \";\" + df['allelicRatio'].astype(str) + \";\" + df['mark']\n",
    "    df = df[['contig', 'start', 'position', 'term', 'refRPKM_ratio', 'altRPKM_ratio']]\n",
    "    return df\n",
    "    \n",
    "os.chdir(\"/Charles/project/ASm6A/Hypothesis_test/ASm6A/\")\n",
    "sig_result_dir, unsig_result_dir = \"sig/\", \"unsig/\"\n",
    "# sig_result_dir, unsig_result_dir = \"sigpv005/\", \"unsigpv005/\" \n",
    "# sig_result_dir, unsig_result_dir = \"sig01/\", \"unsig01/\" ##########\n",
    "os.system(\"mkdir -p %s\" % sig_result_dir)\n",
    "os.system(\"mkdir -p %s\" % unsig_result_dir)\n",
    "file_list = glob.glob(\"*.txt\")\n",
    "for x in file_list:\n",
    "    df_sig, df_unsig = select_sig_and_unsig(x, \"sig\"), select_sig_and_unsig(x, \"unsig\")\n",
    "    print(os.path.basename(x), len(df_sig), len(df_unsig))\n",
    "    res_sig = os.path.join(sig_result_dir, x.replace(\".txt\",\".bed\"))\n",
    "    res_unsig = os.path.join(unsig_result_dir, x.replace(\".txt\",\".bed\"))\n",
    "    df_sig.to_csv(res_sig, sep=\"\\t\", header=False, index=False)\n",
    "    df_unsig.to_csv(res_unsig, sep=\"\\t\", header=False, index=False)\n",
    "    \n",
    "    \n",
    "def overlap_with_m6A(in_dir):\n",
    "    os.chdir(in_dir)\n",
    "    result_dir = \"contained_m6A/\"\n",
    "    os.system(\"mkdir -p %s\" % result_dir)\n",
    "#     m6a_dir = \"/Charles/project/ASm6A/peak_calling/union_three_bedtools/rm_m6Am/\"\n",
    "#     m6a_dir = \"/Charles/project/ASm6A/peak_calling/union_peak_bedtools/rm_m6Am/\"\n",
    "    m6a_dir = \"/Charles/project/ASm6A/peak_calling/merged_peak_MSPC/rm_m6Am/\" # kidney_IP_3_m6A.bed\n",
    "    # m6a_dir = \"/Charles/project/ASm6A/peak_calling/merged_peak_bedtools/rm_m6Am/\"\n",
    "    # m6a_dir = \"/Charles/project/ASm6A/peak_calling/macs2/10e5/rm_m6Am/\"\n",
    "    ##\n",
    "    bed_list = glob.glob(\"%s/*.bed\" % in_dir)\n",
    "    for asm6a in bed_list:\n",
    "        prefix = os.path.basename(asm6a).split(\".\")[0]\n",
    "        m6a = os.path.join(m6a_dir, \"%s_m6A.bed\" % prefix)\n",
    "        res = os.path.join(result_dir, os.path.basename(asm6a))\n",
    "        ###################################################\n",
    "        os.system(\"bedtools intersect -a %s -b %s -wa | sort -u > %s\" % (asm6a, m6a, res))\n",
    "        ####################################################################################\n",
    "        \n",
    "        \n",
    "overlap_with_m6A(\"/Charles/project/ASm6A/Hypothesis_test/ASm6A/sig/\")\n",
    "overlap_with_m6A(\"/Charles/project/ASm6A/Hypothesis_test/ASm6A/unsig/\")\n",
    "# overlap_with_m6A(\"/Charles/project/ASm6A/Hypothesis_test/ASm6A/sigpv005/\")\n",
    "# overlap_with_m6A(\"/Charles/project/ASm6A/Hypothesis_test/ASm6A/unsigpv005/\")\n",
    "\n",
    "\n",
    "def select_highFC(in_dir):\n",
    "    os.chdir(in_dir)\n",
    "    result_dir = \"highFC/\"\n",
    "    os.system(\"mkdir -p %s\" % result_dir)\n",
    "    bed_list = glob.glob(\"%s/*.bed\" % in_dir)\n",
    "    for asm6a in bed_list:\n",
    "        df = pd.read_table(asm6a, header=None)\n",
    "        df.columns = ['contig', 'start', 'position', 'term', 'refRPKM_ratio', 'altRPKM_ratio']\n",
    "        df['mark'] = df['term'].str.split(\";\").str[2]\n",
    "        #####################################################\n",
    "        df = df[((df['mark']==\"ref\")&(df['refRPKM_ratio']>=2)) | ((df['mark']==\"alt\")&(df['altRPKM_ratio']>=2))]\n",
    "        ####################################################\n",
    "        res = os.path.join(result_dir, os.path.basename(asm6a))\n",
    "        df.to_csv(res, sep=\"\\t\", header=False, index=False)\n",
    "        \n",
    "select_highFC(\"/Charles/project/ASm6A/Hypothesis_test/ASm6A/sig/contained_m6A/\")\n",
    "# select_highFC(\"/Charles/project/ASm6A/Hypothesis_test/ASm6A/sigpv005/contained_m6A/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #### difference > 30%\n",
    "### 如果Fisher exact test最终不能用，就用这种方法；\n",
    "\n",
    "################################################################\n",
    "data_dir = \"/Charles/project/ASm6A/Hypothesis_test/ASm6A/sig/contained_m6A/highFC/rename/\"\n",
    "os.chdir(data_dir)\n",
    "result_dir = \"diff30%/\"\n",
    "os.system(\"mkdir -p %s\" % result_dir)\n",
    "##################################################################\n",
    "\n",
    "def calculate_diff(in_txt):\n",
    "    df = pd.read_table(in_txt, header=None)\n",
    "    df.columns = ['contig', 'start', 'position', 'term', 'refRPKM_ratio', 'altRPKM_ratio', 'mark']\n",
    "    df['mini'] = np.where(df['refRPKM_ratio']<df['altRPKM_ratio'], df['refRPKM_ratio'], df['altRPKM_ratio'])\n",
    "    df['diff'] = np.abs(df['refRPKM_ratio'] - df['altRPKM_ratio'])\n",
    "    df_sub = df[df['diff'] > (df['mini']*0.3)]\n",
    "    result_file = os.path.join(result_dir, os.path.basename(in_txt))\n",
    "    df_sub.to_csv(result_file, sep=\"\\t\", header=False, index=False)\n",
    "#     print(in_txt, len(df_sub))\n",
    "\n",
    "txt_list = glob.glob(\"*.bed\")\n",
    "for txt in txt_list:\n",
    "    calculate_diff(txt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Renlab allele m6A analysis\n",
    "###### https://github.com/Jakob666/allele-specificM6A"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### pre-process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### replace refseq accession to ensembl ID\n",
    "\n",
    "def overlap_anno(bed):\n",
    "    result_dir = \"/Charles/project/ASm6A/peak_calling/MeTPeak/forRen_analysis/\"\n",
    "    gene = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/gencode.v19.annotation_Gene.bed\"\n",
    "    res_1 = os.path.join(result_dir, \"%s.bed\" % bed.split(\"/\")[-2])\n",
    "    os.system(\"bedtools intersect -a %s -b %s -wo > %s\" % (bed, gene, res_1))\n",
    "    df = pd.read_table(res_1, header=None)\n",
    "    df.iloc[:,19] = df.iloc[:,19].str.split(\".\").str[0]\n",
    "    df = df.iloc[:, [0,1,2,19,4,5,6,7,8,9,10,11]]\n",
    "    res = res_1\n",
    "    df.to_csv(res, sep=\"\\t\", header=False, index=False)\n",
    "    \n",
    "bed_dir = \"/Charles/project/ASm6A/peak_calling/MeTPeak/\"\n",
    "bed_list = glob.glob(\"%s/*/peak.bed\"%bed_dir)\n",
    "for bed in bed_list:\n",
    "#     print(bed)\n",
    "    overlap_anno(bed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Generate run shell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "vcf_dir = \"/Charles/project/ASm6A/by_ind/SNP_calling/filtered_vcf/forRen_analysis/\"\n",
    "peak_dir = \"/Charles/project/ASm6A/peak_calling/MeTPeak/forRen_analysis/\"\n",
    "gtf = \"/home/mjy/software/allele-specificM6A/Homo_sapiens.GRCh37.87.gtf\"\n",
    "software_dir = \"/home/software/allele-specificM6A/allele-specificM6A-master/\"\n",
    "result_dir = \"/Charles/project/ASm6A/ASm6A_renlab/\"\n",
    "\n",
    "ind_dict = {\n",
    "    \"ind1\": ['brain_IP_1', 'heart_IP_1', 'liver_IP_1'],\n",
    "    \"ind2\":\n",
    "    ['brain_IP_2', 'heart_IP_2', 'placenta_IP_2', 'kidney_IP_2', 'liver_IP_2'],\n",
    "    \"ind3\": ['brain_IP_3', 'heart_IP_3', 'liver_IP_3', 'kidney_IP_3'],\n",
    "    \"ind4\": [\n",
    "        'stomach_IP_4', 'muscle_IP_4', 'lung_IP_4', 'kidney_IP_4',\n",
    "        'placenta_IP_4'\n",
    "    ],\n",
    "    \"ind5\": ['stomach_IP_5', 'muscle_IP_5', 'lung_IP_5'],\n",
    "    \"ind6\": ['placenta_IP_6'],\n",
    "    \"ind7\": [\n",
    "        'Adipose-1-1', 'Adrenal_gland-1-1', 'Aorta-1-1', 'Heart-1-1',\n",
    "        'Skin-1-1', 'Spleen-1-1'\n",
    "    ],\n",
    "    \"ind8\": [\n",
    "        'Lung-2-1', 'Lung-2-4', 'Spleen-2-1', 'Tongue-2-1',\n",
    "        'Urinary_bladder-2-1'\n",
    "    ],\n",
    "    \"ind9\":\n",
    "    ['Appendix-3-2', 'Colon-3-2', 'Esophagus-3-2', 'Muscle-3-2', 'Spleen-3-2'],\n",
    "    \"ind10\": [\n",
    "        'Aorta-4-2', 'Esophagus-4-2', 'Heart-4-2', 'Jejunum-4-2', 'Liver-4-2',\n",
    "        'Lung-4-2', 'Lung-4-4', 'Prostate-4-2', 'Rectum-4-2', 'Skin-4-2',\n",
    "        'Stomach-4-2', 'Testis-4-2', 'Thyroid_gland-4-2', 'Urinary_bladder-4-2'\n",
    "    ],\n",
    "    \"ind11\": [\n",
    "        'Appendix-5-3', 'Brainstem-5-3', 'Cerebellum-5-3', 'Cerebrum-5-3',\n",
    "        'Duodenum-5-3', 'Hypothalamus-5-3', 'Jejunum-5-3', 'Muscle-5-3',\n",
    "        'Rectum-5-3', 'Stomach-5-3', 'Thyroid_gland-5-3', 'Trachea-5-3',\n",
    "        'Urinary_bladder-5-3'\n",
    "    ],\n",
    "    \"ind12\": ['Cerebrum-6-3'],\n",
    "    \"ind13\": ['Cerebellum-7-4'],\n",
    "    \"ind14\": [\n",
    "        'FrontalCortex_1', 'Cerebellum_1', 'Heart_1', 'Liver_1', 'Lung_1',\n",
    "        'Kidney_1', 'Spleen_1'\n",
    "    ],\n",
    "    \"ind15\": [\n",
    "        'FrontalCortex_2', 'Cerebellum_2', 'Heart_2', 'Liver_2', 'Kidney_2',\n",
    "        'Spleen_2', 'Muscle_1'\n",
    "    ],\n",
    "    \"ind16\": [\n",
    "        'FrontalCortex_3', 'Cerebellum_3', 'Heart_3', 'Liver_3', 'Lung_2',\n",
    "        'Muscle_2'\n",
    "    ],\n",
    "    \"ind17\": ['Muscle_3'],\n",
    "    \"ind18\": ['Lung_3', 'Kidney_3', 'Spleen_3']\n",
    "}\n",
    "\n",
    "with open(\"%s/run.sh\" % result_dir, 'w') as fw:\n",
    "    for ind, sample_list in ind_dict.items():\n",
    "        vcf = os.path.join(vcf_dir, \"%s_gatk_het_SNP.vcf\" % ind)\n",
    "        for sample in sample_list:\n",
    "            bed = os.path.join(peak_dir, \"%s.bed\" % sample)\n",
    "            sub_dir = os.path.join(result_dir, sample)\n",
    "            os.system(\"mkdir -p %s\" % sub_dir)\n",
    "            res = os.path.join(sub_dir, \"%s.txt\" % sample)\n",
    "            commond_1 = \"java -jar %s/renlabm6a_allele.jar AsmPeakDetection -g %s -bed %s -vcf %s -o %s -t 5\" % (\n",
    "                software_dir, gtf, bed, vcf, res)\n",
    "            fw.write(commond_1 + \"\\n\")\n",
    "#         os.system(commond_1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### parse results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pickOut_ASm6A(in_file):\n",
    "    #     in_file = \"/Charles/project/ASm6A/ASm6A_renlab/liver_IP_3/liver_IP_3.txt\"\n",
    "    df = pd.read_table(in_file)\n",
    "    df = df[df['q-value'] < 0.05]\n",
    "    asm6a_list = []\n",
    "    for i, values in df.iterrows():\n",
    "        chrom = values['#chr']\n",
    "        peak_start = values['peakStart']\n",
    "        peak_end = values['peakEnd']\n",
    "        term_list = values['majorAlleleNC'].split(\";\")\n",
    "        for term in term_list:\n",
    "            pos, majorAllele = int(term.split(\":\")[0]), term.split(\":\")[1]\n",
    "            start = pos - 1\n",
    "            new_line = \"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\" % (\n",
    "                chrom, str(start), str(pos), majorAllele, chrom,\n",
    "                str(peak_start), str(peak_end))\n",
    "            asm6a_list.append(new_line)\n",
    "    res = in_file.replace(\".txt\", \".bed\")\n",
    "    with open(res, 'w') as fw:\n",
    "        fw.writelines([\"%s\\n\" % x for x in asm6a_list])\n",
    "\n",
    "\n",
    "data_dir = \"/Charles/project/ASm6A/ASm6A_renlab/\"  # liver_IP_3/liver_IP_3.txt\n",
    "name_list = [\n",
    "    os.path.basename(x).replace(\"/\", \"\") for x in glob.glob(\"%s/*\" % data_dir)\n",
    "    if os.path.isdir(x)\n",
    "]\n",
    "sample_list = [\"%s/%s/%s.txt\" % (data_dir, name, name) for name in name_list]\n",
    "# print(sample_list)\n",
    "for sample in sample_list:\n",
    "    pickOut_ASm6A(sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### compare with data-derived methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 这一部分的分析重新做，他们的方法是以peak为单位进行的，所以最后鉴定到的是ASm6A peaks.\n",
    "#### 对于一个 m6A 信号峰，我们假设整个峰值区域存在一个全局的等位基因特异性优势比，\n",
    "#### 该值可理解为修饰峰区域内所有 SNP 位点的等位基因特异性对数优势比的总体平均数(Population average)，\n",
    "#### 则我们可以利用该值估计 m6A 信号峰的等位基因特异性大小。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# asm6a_list = glob.glob(\"/Charles/project/ASm6A/ASm6A_renlab/*/*.bed\")\n",
    "# result_dir = \"/Charles/project/ASm6A/ASm6A_renlab/z_reformat/\"\n",
    "\n",
    "# for asm6a in asm6a_list:\n",
    "#     df = pd.read_table(asm6a, header=None)\n",
    "#     df.iloc[:, 0] = \"chr\" + df.iloc[:, 0].astype(str)\n",
    "#     df.to_csv(os.path.join(result_dir, os.path.basename(asm6a)),\n",
    "#               sep=\"\\t\",\n",
    "#               header=False,\n",
    "#               index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### restrict to same peak regions\n",
    "# asm6a_list = glob.glob(\"/Charles/project/ASm6A/ASm6A_renlab/z_reformat/*.bed\")\n",
    "# peak_dir = \"/Charles/project/ASm6A/peak_calling/merged_peak_MSPC/rm_m6Am/rm_title/rename/\"\n",
    "# result_dir = \"/Charles/project/ASm6A/ASm6A_renlab/z_comparision/restrict_to_samePeak/\"\n",
    "\n",
    "# for asm6a in asm6a_list:\n",
    "#     peak = os.path.join(peak_dir, os.path.basename(asm6a))\n",
    "#     res = os.path.join(result_dir, os.path.basename(asm6a))\n",
    "#     os.system(\"bedtools intersect -a %s -b %s -wa | sort -u > %s\" %\n",
    "#               (asm6a, peak, res))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### overlap with data-derived ASm6A results\n",
    "# renlab_dir = \"/Charles/project/ASm6A/ASm6A_renlab/z_comparision/restrict_to_samePeak/\"\n",
    "# result_dir = \"/Charles/project/ASm6A/ASm6A_renlab/z_comparision/compare_results/\"\n",
    "\n",
    "# # asm6a_dir = \"/Charles/project/ASm6A/ASm6A/\"\n",
    "# asm6a_dir = \"/Charles/project/ASm6A/ASm6A/common_ASm6A/totaltissues/by_sample/\"\n",
    "# sample_list = glob.glob(\"%s/*.bed\" % renlab_dir)\n",
    "# for sample in sample_list:\n",
    "#     asm6a = os.path.join(asm6a_dir, os.path.basename(sample))\n",
    "#     res = os.path.join(result_dir, os.path.basename(sample))\n",
    "#     os.system(\"bedtools intersect -a %s -b %s -wa | sort -u > %s\" %\n",
    "#               (asm6a, sample, res))\n",
    "#     overlap_num = os.popen(\"wc -l %s\" % res).read().split()[0]\n",
    "#     asm6a_num = os.popen(\"wc -l %s\" % asm6a).read().split()[0]\n",
    "#     ren_num = os.popen(\"wc -l %s\" % sample).read().split()[0]\n",
    "#     print(\"%s\\t%s:(%s)\\t%s:(%s)\" %\n",
    "#           (os.path.basename(sample), ren_num,\n",
    "#            round(int(overlap_num) * 100 / int(ren_num), 2), asm6a_num,\n",
    "#            round(int(overlap_num) * 100 / int(asm6a_num), 2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### reanalysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# peak_dir = \"/Charles/project/ASm6A/peak_calling/merged_peak_MSPC/rm_m6Am/rm_title/rename/\"\n",
    "result_dir = \"/Charles/project/ASm6A/ASm6A_renlab/sig_peak/\"\n",
    "ren_dir = \"/Charles/project/ASm6A/ASm6A_renlab/\"\n",
    "bed_list = glob.glob(\"%s/*/*.bed\" % ren_dir)\n",
    "for bed in bed_list:\n",
    "    prefix = bed.split(\"/\")[-2]\n",
    "    df = pd.read_table(bed, sep=\"\\t\", header=None)\n",
    "    df['chr'] = \"chr\" + df_ren.iloc[:,4].astype(str)\n",
    "    df['start'] = df_ren.iloc[:,5].astype(str)\n",
    "    df['end'] =  df_ren.iloc[:,6].astype(str)\n",
    "    df = df[['chr','start','end']].drop_duplicates().dropna()\n",
    "    res = os.path.join(result_dir, \"%s.bed\"%prefix)\n",
    "    df.to_csv(res, sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_dir = \"/Charles/project/ASm6A/ASm6A/sig_peak/\"\n",
    "asm6a_dir = \"/Charles/project/ASm6A/ASm6A/withPeak/\"\n",
    "bed_list = glob.glob(\"%s/*.bed\" % asm6a_dir)\n",
    "for bed in bed_list:\n",
    "    prefix = os.path.basename(bed).split(\".bed\")[0]\n",
    "    df = pd.read_table(bed, sep=\"\\t\", header=None)\n",
    "    df['chr'] = df.iloc[:,6]\n",
    "    df['start'] = df.iloc[:,7].astype(str)\n",
    "    df['end'] = df.iloc[:,8].astype(str)\n",
    "    res = os.path.join(result_dir, \"%s.bed\"%prefix)\n",
    "    df[['chr','start','end']].drop_duplicates().to_csv(res, sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kidney_2.bed 36 1 0.027777777777777776\n",
      "Liver_3.bed 135 16 0.11851851851851852\n",
      "Spleen_3.bed 607 150 0.2471169686985173\n",
      "stomach_IP_4.bed 708 234 0.3305084745762712\n",
      "Appendix-3-2.bed 108 8 0.07407407407407407\n",
      "Liver_1.bed 128 7 0.0546875\n",
      "Cerebellum-5-3.bed 454 110 0.2422907488986784\n",
      "Adrenal_gland-1-1.bed 699 113 0.16165951359084407\n",
      "FrontalCortex_3.bed 409 96 0.23471882640586797\n",
      "liver_IP_3.bed 620 128 0.2064516129032258\n",
      "kidney_IP_4.bed 858 219 0.25524475524475526\n",
      "Cerebellum_3.bed 509 124 0.24361493123772102\n",
      "Aorta-4-2.bed 258 159 0.6162790697674418\n",
      "Muscle_1.bed 70 14 0.2\n",
      "Tongue-2-1.bed 141 20 0.14184397163120568\n",
      "Cerebrum-6-3.bed 168 1 0.005952380952380952\n",
      "Lung_2.bed 396 117 0.29545454545454547\n",
      "Hypothalamus-5-3.bed 163 56 0.34355828220858897\n",
      "Heart-4-2.bed 180 107 0.5944444444444444\n",
      "Cerebrum-5-3.bed 252 71 0.28174603174603174\n",
      "Spleen-3-2.bed 56 9 0.16071428571428573\n",
      "Cerebellum_1.bed 311 30 0.09646302250803858\n",
      "Heart-1-1.bed 640 133 0.2078125\n",
      "Urinary_bladder-2-1.bed 48 5 0.10416666666666667\n",
      "Lung-4-4.bed 16 3 0.1875\n",
      "Rectum-4-2.bed 62 13 0.20967741935483872\n",
      "Kidney_3.bed 348 83 0.23850574712643677\n",
      "Heart_2.bed 256 32 0.125\n",
      "Colon-3-2.bed 161 15 0.09316770186335403\n",
      "heart_IP_2.bed 662 170 0.256797583081571\n",
      "Spleen_1.bed 342 50 0.14619883040935672\n",
      "lung_IP_5.bed 376 79 0.21010638297872342\n",
      "Spleen-2-1.bed 151 37 0.24503311258278146\n",
      "kidney_IP_2.bed 1053 254 0.24121557454890788\n",
      "Skin-1-1.bed 772 161 0.20854922279792745\n",
      "Kidney_1.bed 223 13 0.05829596412556054\n",
      "Brainstem-5-3.bed 299 79 0.26421404682274247\n",
      "Stomach-5-3.bed 137 44 0.32116788321167883\n",
      "Muscle_3.bed 276 20 0.07246376811594203\n",
      "placenta_IP_6.bed 677 64 0.09453471196454949\n",
      "Adipose-1-1.bed 610 141 0.23114754098360657\n",
      "FrontalCortex_2.bed 629 154 0.24483306836248012\n",
      "Lung-2-1.bed 50 12 0.24\n",
      "Muscle-3-2.bed 56 6 0.10714285714285714\n",
      "brain_IP_1.bed 606 94 0.1551155115511551\n",
      "placenta_IP_2.bed 966 245 0.2536231884057971\n",
      "Spleen_2.bed 2 0 0.0\n",
      "Esophagus-4-2.bed 105 36 0.34285714285714286\n",
      "muscle_IP_5.bed 1003 259 0.25822532402791626\n",
      "heart_IP_3.bed 599 143 0.2387312186978297\n",
      "Heart_1.bed 208 22 0.10576923076923077\n",
      "Lung-2-4.bed 90 41 0.45555555555555555\n",
      "heart_IP_1.bed 293 63 0.2150170648464164\n",
      "Spleen-1-1.bed 502 85 0.1693227091633466\n",
      "Liver_2.bed 15 0 0.0\n",
      "Trachea-5-3.bed 204 73 0.35784313725490197\n",
      "Aorta-1-1.bed 350 48 0.13714285714285715\n",
      "liver_IP_2.bed 600 150 0.25\n",
      "Jejunum-5-3.bed 78 12 0.15384615384615385\n",
      "Esophagus-3-2.bed 118 12 0.1016949152542373\n",
      "Cerebellum_2.bed 336 56 0.16666666666666666\n",
      "Jejunum-4-2.bed 84 16 0.19047619047619047\n",
      "Urinary_bladder-5-3.bed 56 17 0.30357142857142855\n",
      "Duodenum-5-3.bed 37 14 0.3783783783783784\n",
      "Thyroid_gland-5-3.bed 155 55 0.3548387096774194\n",
      "Thyroid_gland-4-2.bed 99 72 0.7272727272727273\n",
      "brain_IP_3.bed 790 150 0.189873417721519\n",
      "Appendix-5-3.bed 45 7 0.15555555555555556\n",
      "brain_IP_2.bed 608 137 0.22532894736842105\n",
      "Skin-4-2.bed 45 30 0.6666666666666666\n",
      "Testis-4-2.bed 98 71 0.7244897959183674\n",
      "Lung-4-2.bed 154 111 0.7207792207792207\n",
      "Rectum-5-3.bed 36 7 0.19444444444444445\n",
      "muscle_IP_4.bed 766 226 0.2950391644908616\n",
      "Heart_3.bed 213 60 0.28169014084507044\n",
      "Muscle_2.bed 85 15 0.17647058823529413\n",
      "Stomach-4-2.bed 153 96 0.6274509803921569\n",
      "lung_IP_4.bed 939 254 0.27050053248136313\n",
      "Muscle-5-3.bed 99 70 0.7070707070707071\n",
      "placenta_IP_4.bed 583 133 0.2281303602058319\n",
      "Urinary_bladder-4-2.bed 402 240 0.5970149253731343\n",
      "Lung_1.bed 202 37 0.18316831683168316\n",
      "kidney_IP_3.bed 634 161 0.2539432176656151\n",
      "liver_IP_1.bed 308 30 0.09740259740259741\n",
      "Prostate-4-2.bed 189 93 0.49206349206349204\n",
      "FrontalCortex_1.bed 122 10 0.08196721311475409\n",
      "Cerebellum-7-4.bed 170 2 0.011764705882352941\n",
      "stomach_IP_5.bed 439 118 0.26879271070615035\n",
      "Liver-4-2.bed 115 80 0.6956521739130435\n",
      "Lung_3.bed 107 6 0.056074766355140186\n"
     ]
    }
   ],
   "source": [
    "self_dir = \"/Charles/project/ASm6A/ASm6A/sig_peak/\"\n",
    "ren_dir = \"/Charles/project/ASm6A/ASm6A_renlab/sig_peak/\"\n",
    "bed_list = glob.glob(\"%s/*.bed\" % self_dir)\n",
    "for bed in bed_list:\n",
    "    ren_bed = os.path.join(ren_dir, os.path.basename(bed))\n",
    "    self_num = os.popen(\"wc -l %s\" % bed).read().split()[0]\n",
    "    overlap_num = os.popen(\"bedtools intersect -a %s -b %s -wa | sort -u | wc -l\" % (bed, ren_bed)).read().split()[0]\n",
    "#     print(self_num)\n",
    "#     print(overlap_num)\n",
    "    print(os.path.basename(bed), self_num, overlap_num, int(overlap_num)/ int(self_num))\n",
    "#     break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
